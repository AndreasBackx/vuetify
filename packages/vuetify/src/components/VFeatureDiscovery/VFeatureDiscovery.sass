@import "../../styles/styles.sass"
@import "./_mixins.sass"

// .v-feature-discovery.theme--light
//   +feature-discovery-theme($material-light)
// .v-feature-discovery.theme--dark
//   +feature-discovery-theme($material-dark)

 // Block
.v-feature-discovery
  position: fixed
  z-index: 10

  &--movable
    transition: $primary-transition

// Element
.v-feature-discovery__backdrop
  border-radius: 100%
  height: 100%
  width: 100%
  opacity: 0.96
  transform: scale(0)
  transition: $primary-transition
  display: flex
  justify-content: center
  align-items: center
  +elevation(4)

.v-feature-discovery__highlight
  border-radius: 100%
  position: absolute
  top: calc(50% - (140px / 2))
  left: calc(50% - (140px / 2))
  background: white
  height: 140px
  width: 140px
  display: flex
  justify-content: center
  align-items: center
  transform: scale(0)
  transition: $primary-transition

.v-feature-discovery__highlight:not(.v-feature-discovery__highlight--no-ripple)
  &:before
    background-color: inherit
    content: ''
    border-radius: inherit
    position: absolute
    height: calc(100% + 180px)
    width: calc(100% + 180px)
    transition: $primary-transition
    animation: pulse-animation 1.4s cubic-bezier(0.24, 0, 0.38, 1) infinite
    pointer-events: none

.v-feature-discovery__content
  width: 60%
  height: 60%
  display: flex
  flex-direction: column
  justify-content: space-around
  align-items: center

// Modifier
.v-feature-discovery--active
  .v-feature-discovery__backdrop
    transform: scale(1)

  .v-feature-discovery__highlight
    transform: scale(1)

.v-feature-discovery--active
  .v-feature-discovery__highlight:not(.v-feature-discovery__highlight--no-ripple)
    animation: highlight-pulse-animation 1.4s cubic-bezier(0.24, 0, 0.38, 1) infinite

.v-feature-discovery--flat .v-feature-discovery__backdrop
  +elevation(0)

@keyframes highlight-pulse-animation
  from
    transform: scale(1)

  75%
    transform: scale(1.05)

  to
    transform: scale(1)

@keyframes pulse-animation
  from
    opacity: 1
    transform: scale(.2)

  to
    opacity: 0
    transform: scale(1)
