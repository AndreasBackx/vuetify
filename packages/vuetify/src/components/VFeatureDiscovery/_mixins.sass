@mixin feature-discovery-theme($material)
  color: map-deep-get($material, 'text', 'primary')
  background-color: map-get($material, 'app-bar')
