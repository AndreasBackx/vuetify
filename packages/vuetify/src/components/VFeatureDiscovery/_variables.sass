$feature-discovery-transition: $primary-transition, top 0, left 0
