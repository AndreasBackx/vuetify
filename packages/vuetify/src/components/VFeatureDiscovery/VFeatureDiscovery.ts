// Styles
import './VFeatureDiscovery.sass'

// Mixins
import Colorable from '../../mixins/colorable'
import Elevatable from '../../mixins/elevatable'
import Toggleable from '../../mixins/toggleable'
import Themeable from '../../mixins/themeable'

// Types
import { VNode } from 'vue'
import mixins from '../../util/mixins'

// Directives
import ClickOutside from '../../directives/click-outside'
import { convertToUnit, keyCodes } from '../../util/helpers'

const doubledSqrt2 = 2.8284

export default mixins(
  Colorable,
  Elevatable,
  Toggleable,
  Themeable
/* @vue/component */
).extend({
  name: 'v-feature-discovery',

  directives: { ClickOutside },

  props: {
    persistent: Boolean,
    flat: Boolean,
    color: {
      type: String,
      default: 'primary'
    },
    highlightColor: {
      type: String,
      default: 'white'
    },
    size: {
      default: 700,
      type: [Number, String],
      validator: (v: string | number) => !isNaN(parseInt(v))
    },
    target: {
      default: null as string | Element | null
    },
    value: {
      default: true
    },
    noRipple: {
      default: false,
      type: Boolean
    }
  },

  data: () => ({
    rect: {
      top: 0,
      bottom: 0,
      left: 0,
      right: 0,
      height: 0,
      width: 0
    },
    targetEl: null as Element | null,
    oldZIndex: 0,
    movable: true
  }),

  computed: {
    classes (): object {
      return {
        'v-feature-discovery--flat': this.flat,
        'v-feature-discovery--active': this.isActive,
        'v-feature-discovery--movable': this.movable,
        'v-feature-discovery--lr-shifted': this.leftShift !== 0,
        ...this.themeClasses
      }
    },
    computedLeft (): number {
      return this.rect.left - (this.computedSize / 2) + (this.rect.width / 2) + this.leftShift
    },
    computedTop (): number {
      return this.rect.top - (this.computedSize / 2) + (this.rect.height / 2) + this.topShift
    },
    computedSize (): number {
      return parseInt(this.size)
    },
    styles (): object {
      return {
        left: convertToUnit(this.computedLeft),
        top: convertToUnit(this.computedTop),
        height: convertToUnit(this.computedSize),
        width: convertToUnit(this.computedSize),
        'pointer-events': this.isActive ? 'auto' : 'none'
      }
    },
    highlightSize (): number {
      return Math.sqrt(2 * (Math.max(this.rect.width, this.rect.height) + 25) ** 2)
    },
    leftShift (): number {
      if ((this.rect.left - (this.computedSize / 2) + (this.rect.width / 2)) < 0) return this.computedSize / doubledSqrt2 * 0.75
      if ((window.innerWidth - this.rect.right - (this.computedSize / 2) + (this.rect.width / 2)) < 0) {
        return -this.computedSize / doubledSqrt2 * 0.75
      }
      return 0
    },
    topShift (): number {
      if ((this.rect.top - (this.computedSize / 2) + (this.rect.height / 2)) < 0) return this.computedSize / doubledSqrt2 * 0.75
      if ((window.innerHeight - this.rect.bottom - (this.computedSize / 2) + (this.rect.height / 2)) < 0) {
        return -this.computedSize / doubledSqrt2 * 0.75
      }
      return 0
    },
    backdropStyle (): object {
      let justify = 'center'
      let paddingL
      let paddingR

      if (this.leftShift > 0) {
        justify = 'flex-end'
        paddingR = '15%'
      }
      if (this.leftShift < 0) {
        justify = 'flex-start'
        paddingL = '15%'
      }

      return {
        'justify-content': justify,
        'padding-left': paddingL,
        'padding-right': paddingR,
        'transform-origin': `calc(50% - ${this.leftShift}px) calc(50% - ${this.topShift}px)`
      }
    },
    contentStyle (): object {
      let justify = 'space-between'
      if (this.topShift > 0) justify = 'flex-end'
      if (this.topShift < 0) justify = 'flex-start'

      return {
        'justify-content': justify
      }
    }
  },

  watch: {
    target () {
      this.updateTarget()
    },
    isActive (val: boolean) {
      if (!this.targetEl) return

      if (val) this.updateRect();
      (this.targetEl as any).style.zIndex = val ? 11 : this.oldZIndex
    }
  },

  mounted () {
    this.updateTarget()

    window.addEventListener('scroll', this.updateRect)
    window.addEventListener('resize', this.updateRect)
    window.addEventListener('keyup', this.keyPress)
  },

  methods: {
    keyPress (e: KeyboardEvent) {
      if (this.closeConditional() && e.keyCode === keyCodes.esc) this.isActive = false
    },
    updateTarget () {
      if (this.targetEl) (this.targetEl as any).style.zIndex = this.oldZIndex

      if (this.target instanceof Element) this.targetEl = this.target
      else this.targetEl = document.querySelector(this.target)

      if (!this.targetEl) return

      this.rect = this.targetEl.getBoundingClientRect()
      this.oldZIndex = Number(getComputedStyle(this.targetEl).zIndex) || 0
      /* istanbul ignore next */
      if (this.isActive) (this.targetEl as any).style.zIndex = 11
    },
    updateRect () {
      if (!this.targetEl || !this.isActive) return

      this.movable = false
      this.rect = this.targetEl.getBoundingClientRect()

      setTimeout(() => (this.movable = true))
    },
    closeConditional (): boolean {
      return !this.persistent && this.isActive
    },
    genBackdrop (): VNode {
      return this.$createElement('div', this.setBackgroundColor(this.color, {
        staticClass: 'v-feature-discovery__backdrop',
        class: this.elevationClasses,
        style: this.backdropStyle
      }), [ this.genContent() ])
    },
    genContent (): VNode {
      return this.$createElement('div', {
        staticClass: 'v-feature-discovery__content',
        style: this.contentStyle
      }, [
        this.genWrapper(),
        this.genActions()
      ])
    },
    genActions (): VNode {
      return this.$createElement('div', {
        staticClass: 'v-feature-discovery__actions'
      }, this.$scopedSlots.actions ? this.$scopedSlots.actions({
        close: () => {
          if (this.closeConditional()) this.isActive = false
        }
      }) : this.$slots.actions)
    },
    genWrapper (): VNode {
      return this.$createElement('div', {
        staticClass: 'v-feature-discovery__wrapper'
      }, [
        this.$createElement('div', {
          staticClass: 'v-feature-discovery__title'
        }, this.$slots.title),
        this.$createElement('div', {
          staticClass: 'v-feature-discovery__text'
        }, this.$slots.default)
      ])
    },
    genChildren (): VNode[] {
      return [
        this.genBackdrop(),
        this.genHighlight()
      ]
    },
    genHighlight (): VNode {
      return this.$createElement('div', this.setBackgroundColor(this.highlightColor, {
        staticClass: 'v-feature-discovery__highlight',
        class: {
          'v-feature-discovery__highlight--no-ripple': this.noRipple
        },
        style: {
          top: `calc(50% - (${this.highlightSize}px / 2) - ${this.topShift}px)`,
          left: `calc(50% - (${this.highlightSize}px / 2) - ${this.leftShift}px)`,
          height: `${this.highlightSize}px`,
          width: `${this.highlightSize}px`
        }
      }))
    }
  },

  render (h): VNode {
    const data = {
      staticClass: 'v-feature-discovery',
      directives: [
        {
          name: 'click-outside',
          value: () => /* istanbul ignore next */ (this.isActive = false),
          args: {
            closeConditional: this.closeConditional
          }
        }
      ],
      class: this.classes,
      style: this.styles
    }

    return h('div', data, this.genChildren())
  }
})
