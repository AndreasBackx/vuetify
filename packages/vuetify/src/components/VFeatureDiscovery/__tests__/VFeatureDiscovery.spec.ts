// Components
import VFeatureDiscovery from '../VFeatureDiscovery'

// Utilities
import {
  mount,
  Wrapper,
  MountOptions
} from '@vue/test-utils'
import { ExtractVue } from '../../../util/mixins'

describe('FeatureDiscovery.ts', () => {
  type Instance = ExtractVue<typeof VFeatureDiscovery>
  let mountFunction: (options?: object) => Wrapper<Instance>

  beforeEach(() => {
    mountFunction = (options = {} as MountOptions<Instance>) => {
      return mount(VFeatureDiscovery, options)
    }
  })

  it('should get closeConditional', () => {
    const wrapper = mountFunction()

    expect(wrapper.vm.closeConditional()).toBeTruthy()
    wrapper.setProps({
      persistent: true
    })
    expect(wrapper.vm.closeConditional()).toBeFalsy()
    wrapper.setProps({
      persistent: false
    })
    wrapper.setData({
      isActive: false
    })
    expect(wrapper.vm.closeConditional()).toBeFalsy()
  })

  it('should render correctly', () => {
    const wrapper = mountFunction()

    expect(wrapper.classes('v-feature-discovery')).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly when flat', () => {
    const wrapper = mountFunction({
      propsData: {
        flat: true
      }
    })

    expect(wrapper.classes('v-feature-discovery--flat')).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should toggle', () => {
    const wrapper = mountFunction({
      propsData: {
        target: document.body
      }
    })

    expect(wrapper.classes('v-feature-discovery--active')).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.vm.isActive = false

    expect(wrapper.classes('v-feature-discovery--active')).toBeFalsy()
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.vm.isActive = true
    wrapper.vm.isActive = false
  })

  it('should work with scroll properly', async () => {
    const fn = jest.fn()

    const wrapper = mountFunction()

    wrapper.vm.$watch('movable', fn)

    expect(fn).toHaveBeenCalledTimes(0)
    wrapper.vm.updateRect()
    wrapper.setData({
      targetEl: {
        getBoundingClientRect: () => ({
          top: 1,
          bottom: 2,
          left: 3,
          right: 4,
          height: 5,
          width: 6
        }),
        style: {}
      }
    })
    expect(fn).toHaveBeenCalledTimes(0)
    wrapper.vm.updateRect()
    await wrapper.vm.$nextTick()
    expect(wrapper.vm.rect).toEqual({
      top: 1,
      bottom: 2,
      left: 3,
      right: 4,
      height: 5,
      width: 6
    })
    expect(fn).toHaveBeenCalledWith(false, true)
  })

  it('should update targetEl when target prop updates', () => {
    const wrapper = mountFunction()

    const spy = jest.spyOn(wrapper.vm, 'updateTarget')

    expect(spy).toHaveBeenCalledTimes(0)
    wrapper.setProps({
      target: '#asd'
    })
    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('should close on esc', () => {
    let close = false
    const wrapper = mountFunction({
      methods: {
        closeConditional: () => close
      }
    })

    wrapper.vm.keyPress({ keyCode: 27 } as any)
    expect(wrapper.vm.isActive).toBeTruthy()

    close = true
    wrapper.vm.keyPress({ keyCode: 27 } as any)
    expect(wrapper.vm.isActive).toBeFalsy()
  })

  it('should close on close method of actions scoped slot', () => {
    let close = false
    const wrapper = mountFunction({
      methods: {
        closeConditional: () => close
      },
      scopedSlots: {
        actions: `
          <template slot-scope="{ close }">
            <button id="close" @click="close">Close</button>
          </template>
        `
      }
    })

    const closeButton = wrapper.find('#close')

    closeButton.trigger('click')
    expect(wrapper.vm.isActive).toBeTruthy()

    close = true
    closeButton.trigger('click')
    expect(wrapper.vm.isActive).toBeFalsy()
  })

  it('should compute base shift', async () => {
    const wrapper1 = mountFunction({
      computed: {
        computedSize: () => 500
      }
    })

    expect(wrapper1.vm.baseShift).toBeCloseTo(132.58, 1)

    const wrapper2 = mountFunction({
      computed: {
        computedSize: () => 100
      }
    })

    expect(wrapper2.vm.baseShift).toBeCloseTo(26.51, 1)
  })

  it('should compute left shift', async () => {
    const wrapper = mountFunction({
      computed: {
        computedSize: () => 500
      }
    })

    wrapper.setData({
      rect: {
        left: -100,
        right: 0,
        width: 100
      }
    })
    expect(wrapper.vm.leftShift).toBeCloseTo(132.58, 1)

    wrapper.setData({
      rect: {
        left: 1024,
        right: 1124,
        width: 100
      }
    })
    expect(wrapper.vm.leftShift).toBeCloseTo(-132.58, 1)

    wrapper.setData({
      rect: {
        left: 500,
        right: 600,
        width: 100
      }
    })
    expect(wrapper.vm.leftShift).toBe(0)
  })

  it('should compute top shift', async () => {
    const wrapper = mountFunction({
      computed: {
        computedSize: () => 500
      }
    })

    wrapper.setData({
      rect: {
        top: -100,
        bottom: 0,
        height: 100
      }
    })
    expect(wrapper.vm.topShift).toBeCloseTo(132.58, 1)

    wrapper.setData({
      rect: {
        top: 768,
        bottom: 868,
        height: 100
      }
    })
    expect(wrapper.vm.topShift).toBeCloseTo(-132.58, 1)

    wrapper.setData({
      rect: {
        top: 400,
        bottom: 500,
        height: 100
      }
    })
    expect(wrapper.vm.topShift).toBe(0)
  })

  it('should compute rect size', async () => {
    const wrapper = mountFunction()

    wrapper.setData({
      rect: {
        width: 200,
        height: 100
      }
    })
    expect(wrapper.vm.rectSize).toBe(200)

    wrapper.setData({
      rect: {
        width: 200,
        height: 200
      }
    })
    expect(wrapper.vm.rectSize).toBe(200)

    wrapper.setData({
      rect: {
        width: 200.1,
        height: 100
      }
    })
    expect(wrapper.vm.rectSize).toBe(200.1)
  })

  it('should compute highlight padding', async () => {
    const wrapper = mountFunction()

    wrapper.setData({
      rect: {
        width: 200,
        height: 100
      }
    })
    expect(wrapper.vm.highlightPadding).toBeCloseTo(57.14, 1)
  })
})
