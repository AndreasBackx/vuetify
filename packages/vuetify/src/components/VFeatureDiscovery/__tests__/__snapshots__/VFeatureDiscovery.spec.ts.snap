// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureDiscovery.ts should render correctly 1`] = `

<div class="v-feature-discovery v-feature-discovery--active v-feature-discovery--movable v-feature-discovery--lr-shifted theme--light"
     style="left: -102.51025314665534px; top: -102.51025314665534px; height: 700px; width: 700px; pointer-events: auto;"
>
  <div class="v-feature-discovery__backdrop primary"
       style="padding-right: 10%;"
  >
    <div class="v-feature-discovery__content">
      <div class="v-feature-discovery__wrapper">
        <div class="v-feature-discovery__title">
        </div>
        <div class="v-feature-discovery__text">
        </div>
      </div>
      <div class="v-feature-discovery__actions">
      </div>
    </div>
  </div>
  <div class="v-feature-discovery__highlight white"
       style="height: 35.35533905932738px; width: 35.35533905932738px;"
  >
  </div>
</div>

`;

exports[`FeatureDiscovery.ts should render correctly when flat 1`] = `

<div class="v-feature-discovery v-feature-discovery--flat v-feature-discovery--active v-feature-discovery--movable v-feature-discovery--lr-shifted theme--light"
     style="left: -102.51025314665534px; top: -102.51025314665534px; height: 700px; width: 700px; pointer-events: auto;"
>
  <div class="v-feature-discovery__backdrop primary"
       style="padding-right: 10%;"
  >
    <div class="v-feature-discovery__content">
      <div class="v-feature-discovery__wrapper">
        <div class="v-feature-discovery__title">
        </div>
        <div class="v-feature-discovery__text">
        </div>
      </div>
      <div class="v-feature-discovery__actions">
      </div>
    </div>
  </div>
  <div class="v-feature-discovery__highlight white"
       style="height: 35.35533905932738px; width: 35.35533905932738px;"
  >
  </div>
</div>

`;

exports[`FeatureDiscovery.ts should toggle 1`] = `

<div class="v-feature-discovery v-feature-discovery--active v-feature-discovery--movable v-feature-discovery--lr-shifted theme--light"
     style="left: -102.51025314665534px; top: -102.51025314665534px; height: 700px; width: 700px; pointer-events: auto;"
>
  <div class="v-feature-discovery__backdrop primary"
       style="padding-right: 10%;"
  >
    <div class="v-feature-discovery__content">
      <div class="v-feature-discovery__wrapper">
        <div class="v-feature-discovery__title">
        </div>
        <div class="v-feature-discovery__text">
        </div>
      </div>
      <div class="v-feature-discovery__actions">
      </div>
    </div>
  </div>
  <div class="v-feature-discovery__highlight white"
       style="height: 35.35533905932738px; width: 35.35533905932738px;"
  >
  </div>
</div>

`;

exports[`FeatureDiscovery.ts should toggle 2`] = `

<div class="v-feature-discovery v-feature-discovery--movable v-feature-discovery--lr-shifted theme--light"
     style="left: -102.51025314665534px; top: -102.51025314665534px; height: 700px; width: 700px; pointer-events: none;"
>
  <div class="v-feature-discovery__backdrop primary"
       style="padding-right: 10%;"
  >
    <div class="v-feature-discovery__content">
      <div class="v-feature-discovery__wrapper">
        <div class="v-feature-discovery__title">
        </div>
        <div class="v-feature-discovery__text">
        </div>
      </div>
      <div class="v-feature-discovery__actions">
      </div>
    </div>
  </div>
  <div class="v-feature-discovery__highlight white"
       style="height: 35.35533905932738px; width: 35.35533905932738px;"
  >
  </div>
</div>

`;
